### PR TITLE
Add multi_precision for adagrad op

### PR DESCRIPTION
**Pull request by AnnaTrainingG**
### PR types

New features 
### PR changes

APIs
### Describe

给 adagrad api 新增multi_precision 参数用于AMP O2训练

### CODE REVIEW

**Review by Shixiaowei02 — APPROVED** [review]

**Review by jiweibo — APPROVED** [review]
LGTM for `dispensable inout` and `default attr`

File: paddle/phi/kernels/gpu/adagrad_kernel.cu [hunk 1]
```diff
@@ -13,16 +13,91 @@
...
+
+  for (int i = idx; i < num; i += blockDim.x * gridDim.x) {
+    MT grad_data = static_cast<MT>(grad[i]);
+    MT moment_out_data = static_cast<MT>(moment[i]) + grad_data * grad_data;
+    moment_out[i] = static_cast<MT>(moment_out_data);
+    auto in = master_param_out ? master_param[i] : static_cast<MT>(param[i]);
+    MT param_out_data =
+        in - (lr_data * grad_data) / (sqrt(moment_out_data) + epsilon);
+
+    param_out[i] = static_cast<MT>(param_out_data);
```
Comment (liudongxue01) [inline paddle/phi/kernels/gpu/adagrad_kernel.cu:50]: 这里是不是应该改成  param_out[i] = static_cast< T >(param_out_data);

File: python/paddle/optimizer/adagrad.py [hunk 2]
```diff
@@ -126,19 +130,91 @@ def __init__(
...
     def _create_accumulators(self, block, parameters):
         assert isinstance(block, framework.Block)
 
         if isinstance(parameters, dict):
             parameters = self._update_param_group(parameters)
 
         for p in parameters:
+            if self._multi_precision and p.dtype == core.VarDesc.VarType.FP16:
+                master_p = self._create_master_weight(p)
+                self._add_accumulator(self._moment_acc_str, master_p)
```
Comment (liudongxue01) [inline python/paddle/optimizer/adagrad.py:208]: 此处是不是缺少fill_value=self.initial_accumulator_value？

File: paddle/phi/kernels/gpu/adagrad_kernel.cu (hunk 1 above)
Comment (AnnaTrainingG) [reply paddle/phi/kernels/gpu/adagrad_kernel.cu:50]: https://github.com/PaddlePaddle/Paddle/pull/51790 修复PR 已提交

File: python/paddle/optimizer/adagrad.py (hunk 2 above)
Comment (AnnaTrainingG) [reply python/paddle/optimizer/adagrad.py:208]: https://github.com/PaddlePaddle/Paddle/pull/51790 修复PR已经提交

**Pull request merged**